File: Data Structures & Algorithms/validate-parentheses/submission-1.js
class Solution {
    /**
     * @param {string} s
     * @return {boolean}
     */
    isValid(s) {
        let sta = [];

        for (let i = 0; i < s.length; i++) {
            if (s[i] == '(' || s[i] == '{' || s[i] == '[') {
                sta.push(s[i]);
            } else {
                switch (s[i]) {
                    case ')':
                        if (sta.length > 0 && sta[sta.length - 1] == '(') {
                            sta.pop();
                        } else {
                            return false;
                        }
                        break;

                    case '}':
                        if (sta.length > 0 && sta[sta.length - 1] == '{') {
                            sta.pop();
                        } else {
                            return false;
                        }
                        break;
        
                    case ']':
                        if (sta.length > 0 && sta[sta.length - 1] == '[') {
                            sta.pop();
                        } else {
                            return false;
                        }
                        break;

                    default:
                        return false;
                }
            }
        }

        return sta.length === 0;
    }
}

File: Data Structures & Algorithms/subsets/submission-1.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = []
        subset = []

        def dfs(n):
            # base case: n is out of array nums bounds
            if n >= len(nums):
                result.append(subset.copy())
                return
            
            # explore the depth of the branch including nums[n]
            subset.append(nums[n])
            dfs(n + 1)

            # explore the depth of the branch excluding nums[n]
            subset.pop()
            dfs(n + 1)
        
        dfs(0)
        return result
        